libdnf/sack/advisory.cpp

```diff
@@ -151,9 +151,66 @@ std::vector<AdvisoryModule> Advisory::getModules() const
         moduleList.emplace_back(sack, advisory, name, stream, version, context, arch);
     }
     dataiterator_free(&di);
+
     return moduleList;
 }
 
+std::vector<AdvisoryPkg>
+Advisory::getApplicablePackages() const
+{
+    std::vector<AdvisoryPkg> pkgList;
+
+    Dataiterator di;
+    Dataiterator di_inner;
+    Pool *pool = dnf_sack_get_pool(sack);
+
+    dataiterator_init(&di, pool, 0, advisory, UPDATE_COLLECTIONLIST, 0, 0);
+    while (dataiterator_step(&di)) {
+        dataiterator_setpos(&di);
+
+        std::vector<AdvisoryModule> moduleList;
+        const char * filename = nullptr;
+
+        dataiterator_init(&di_inner, pool, 0, SOLVID_POS, UPDATE_MODULE, 0, 0);
+        while (dataiterator_step(&di_inner)) {
+            dataiterator_setpos(&di_inner);
+            Id name = pool_lookup_id(pool, SOLVID_POS, UPDATE_MODULE_NAME);
+            Id stream = pool_lookup_id(pool, SOLVID_POS, UPDATE_MODULE_STREAM);
+            Id version = pool_lookup_id(pool, SOLVID_POS, UPDATE_MODULE_VERSION);
+            Id context = pool_lookup_id(pool, SOLVID_POS, UPDATE_MODULE_CONTEXT);
+            Id arch = pool_lookup_id(pool, SOLVID_POS, UPDATE_MODULE_ARCH);
+            moduleList.emplace_back(sack, advisory, name, stream, version, context, arch);
+        }
+        dataiterator_free(&di_inner);
+
+        bool isModuleCollectionApplicable = true;
+        for (auto & moduleAdvisory: moduleList) {
+            if (!moduleAdvisory.isApplicable()) {
+                isModuleCollectionApplicable = false;
+                break;
+            }
+        }
```

Comment:
Somehow I believe that the logic is not the proper one. The logic says that when any module in collection is not applicable then the collection is not applicable. What about to shorten it like:
```
        bool isModuleCollectionApplicable = true;
        dataiterator_init(&di_inner, pool, 0, SOLVID_POS, UPDATE_MODULE, 0, 0);
        while (dataiterator_step(&di_inner)) {
            isModuleCollectionApplicable = false;
            dataiterator_setpos(&di_inner);
            Id name = pool_lookup_id(pool, SOLVID_POS, UPDATE_MODULE_NAME);
            Id stream = pool_lookup_id(pool, SOLVID_POS, UPDATE_MODULE_STREAM);
            Id version = pool_lookup_id(pool, SOLVID_POS, UPDATE_MODULE_VERSION);
            Id context = pool_lookup_id(pool, SOLVID_POS, UPDATE_MODULE_CONTEXT);
            Id arch = pool_lookup_id(pool, SOLVID_POS, UPDATE_MODULE_ARCH);
            AdvisoryModule moduleAdvisory(sack, advisory, name, stream, version, context, arch);
            if (moduleAdvisory.isApplicable()) {
                isModuleCollectionApplicable = true;
                break;
            }
        }
        dataiterator_free(&di_inner);
```

Reply:
Fixed, thank you.

Reply:
Ok, I have fixed the logic as well now. Only one applicable module is sufficient to make a collection applicable.